File: labs/M20-monitoring/solution/alert_engine.py
"""
M20 Lab — Monitoring: Alert Engine (Solution)
==============================================
Complete implementation of the alert engine with default rules
for error rate, latency, cost, tool failures, and staleness.

Usage:
    python alert_engine.py
"""

import time
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Any, Optional


# =============================================================================
# DATA CLASSES
# =============================================================================

@dataclass
class AlertRule:
    """A rule that defines when an alert should fire."""
    name: str
    condition_fn: Callable[[Dict], bool]
    severity: str  # "critical" | "warning" | "info"
    message_template: str
    value_fn: Optional[Callable[[Dict], Any]] = None


@dataclass
class Alert:
    """A triggered alert."""
    rule_name: str
    severity: str
    message: str
    timestamp: float
    metric_value: Any


# =============================================================================
# ALERT ENGINE
# =============================================================================

class AlertEngine:
    """
    Evaluates metrics against registered rules and produces alerts.
    """

    def __init__(self):
        self.rules: List[AlertRule] = []
        self._setup_default_rules()

    def _setup_default_rules(self) -> None:
        """Register the default set of alert rules."""

        # Rule 1: High error rate (>5% = critical)
        self.add_rule(AlertRule(
            name="high_error_rate",
            condition_fn=lambda m: m.get("error_rate", 0) > 5,
            severity="critical",
            message_template="Error rate is {value:.1f}% (threshold: 5%)",
            value_fn=lambda m: m.get("error_rate", 0),
        ))

        # Rule 2: High latency p95 (>10,000ms = warning)
        self.add_rule(AlertRule(
            name="high_latency_p95",
            condition_fn=lambda m: m.get("latency", {}).get("p95", 0) > 10000,
            severity="warning",
            message_template="P95 latency is {value:.0f}ms (threshold: 10,000ms)",
            value_fn=lambda m: m.get("latency", {}).get("p95", 0),
        ))

        # Rule 3: High cost (>$1/hr = warning)
        self.add_rule(AlertRule(
            name="high_cost",
            condition_fn=lambda m: m.get("tokens", {}).get("cost_per_hour", 0) > 1.0,
            severity="warning",
            message_template="Cost rate is ${value:.2f}/hr (threshold: $1.00/hr)",
            value_fn=lambda m: m.get("tokens", {}).get("cost_per_hour", 0),
        ))

        # Rule 4: Tool failure rate (any tool >10% = critical)
        def check_tool_failure(m):
            tools = m.get("tools", {})
            for tool_name, stats in tools.items():
                if stats.get("failure_rate", 0) > 10:
                    return True
            return False

        def get_max_tool_failure(m):
            tools = m.get("tools", {})
            max_rate = 0
            for tool_name, stats in tools.items():
                rate = stats.get("failure_rate", 0)
                if rate > max_rate:
                    max_rate = rate
            return max_rate

        self.add_rule(AlertRule(
            name="tool_failure_rate",
            condition_fn=check_tool_failure,
            severity="critical",
            message_template="Tool failure rate is {value:.1f}% (threshold: 10%)",
            value_fn=get_max_tool_failure,
        ))

        # Rule 5: Stale — no requests for 5+ minutes
        self.add_rule(AlertRule(
            name="stale_no_requests",
            condition_fn=lambda m: (
                m.get("throughput_rpm", 0) == 0
                and m.get("uptime_seconds", 0) > 300
            ),
            severity="info",
            message_template="No requests in last 5 minutes (uptime: {value:.0f}s)",
            value_fn=lambda m: m.get("uptime_seconds", 0),
        ))

    def add_rule(self, rule: AlertRule) -> None:
        """Register a new alert rule."""
        self.rules.append(rule)

    def evaluate(self, metrics: Dict) -> List[Alert]:
        """Evaluate all rules against the given metrics."""
        triggered: List[Alert] = []

        for rule in self.rules:
            try:
                if rule.condition_fn(metrics):
                    value = rule.value_fn(metrics) if rule.value_fn else None
                    message = rule.message_template.format(value=value) if value is not None else rule.message_template
                    triggered.append(Alert(
                        rule_name=rule.name,
                        severity=rule.severity,
                        message=message,
                        timestamp=time.time(),
                        metric_value=value,
                    ))
            except Exception as e:
                # Don't let one bad rule stop evaluation of others
                pass

        return triggered


# =============================================================================
# SELF-TEST
# =============================================================================

def self_test():
    """Test alert engine with metrics that trigger various alerts."""
    print("=" * 60)
    print("M20 Alert Engine — Self-Test")
    print("=" * 60)

    engine = AlertEngine()

    # --- Test 1: Healthy metrics ---
    healthy_metrics = {
        "request_count": 500,
        "error_rate": 1.2,
        "latency": {"p50": 200, "p75": 400, "p90": 800, "p95": 1500, "p99": 3000},
        "tokens": {
            "total_input": 500000,
            "total_output": 250000,
            "cost_estimate": 0.50,
            "cost_per_hour": 0.25,
        },
        "tools": {
            "search_filings": {"calls": 300, "failures": 5, "failure_rate": 1.7, "avg_duration_ms": 45},
            "get_details": {"calls": 200, "failures": 2, "failure_rate": 1.0, "avg_duration_ms": 12},
        },
        "throughput_rpm": 8.3,
        "uptime_seconds": 3600,
    }

    alerts = engine.evaluate(healthy_metrics)
    print(f"\nHealthy metrics -> {len(alerts)} alerts")
    assert len(alerts) == 0, f"Expected 0 alerts for healthy metrics, got {len(alerts)}"
    print("  ✅ No alerts — PASS")

    # --- Test 2: Degraded metrics ---
    degraded_metrics = {
        "request_count": 500,
        "error_rate": 8.5,
        "latency": {"p50": 2000, "p75": 5000, "p90": 9000, "p95": 12000, "p99": 18000},
        "tokens": {
            "total_input": 5000000,
            "total_output": 2500000,
            "cost_estimate": 52.50,
            "cost_per_hour": 2.10,
        },
        "tools": {
            "search_filings": {"calls": 300, "failures": 45, "failure_rate": 15.0, "avg_duration_ms": 145},
            "get_details": {"calls": 200, "failures": 2, "failure_rate": 1.0, "avg_duration_ms": 12},
        },
        "throughput_rpm": 8.3,
        "uptime_seconds": 3600,
    }

    alerts = engine.evaluate(degraded_metrics)
    print(f"\nDegraded metrics -> {len(alerts)} alerts")
    for alert in alerts:
        severity_icon = {"critical": "🔴", "warning": "🟡", "info": "🔵"}[alert.severity]
        print(f"  {severity_icon} [{alert.severity.upper()}] {alert.rule_name}: {alert.message}")

    assert len(alerts) >= 3, f"Expected at least 3 alerts, got {len(alerts)}"
    severities = [a.severity for a in alerts]
    assert "critical" in severities, "Expected at least one critical alert"
    print("  ✅ Multiple alerts with correct severities — PASS")

    # --- Test 3: Stale metrics ---
    stale_metrics = {
        "request_count": 0,
        "error_rate": 0,
        "latency": {"p50": 0, "p75": 0, "p90": 0, "p95": 0, "p99": 0},
        "tokens": {
            "total_input": 0,
            "total_output": 0,
            "cost_estimate": 0,
            "cost_per_hour": 0,
        },
        "tools": {},
        "throughput_rpm": 0,
        "uptime_seconds": 600,
    }

    alerts = engine.evaluate(stale_metrics)
    print(f"\nStale metrics -> {len(alerts)} alerts")
    for alert in alerts:
        severity_icon = {"critical": "🔴", "warning": "🟡", "info": "🔵"}[alert.severity]
        print(f"  {severity_icon} [{alert.severity.upper()}] {alert.rule_name}: {alert.message}")

    stale_alerts = [a for a in alerts if a.rule_name == "stale_no_requests"]
    assert len(stale_alerts) >= 1, "Expected stale alert"
    print("  ✅ Stale detection works — PASS")

    print(f"\n{'=' * 60}")
    print("Self-test complete!")
    print(f"{'=' * 60}")


if __name__ == "__main__":
    self_test()

File: labs/M20-monitoring/solution/alert_engine.js
/**
 * M20 Lab — Monitoring: Alert Engine (Solution)
 * ==============================================
 * Complete implementation of the alert engine.
 *
 * Usage:
 *     node alert_engine.js
 */

class AlertRule {
  constructor(name, conditionFn, severity, messageTemplate, valueFn = null) {
    this.name = name;
    this.conditionFn = conditionFn;
    this.severity = severity;
    this.messageTemplate = messageTemplate;
    this.valueFn = valueFn;
  }
}

class Alert {
  constructor(ruleName, severity, message, timestamp, metricValue) {
    this.ruleName = ruleName;
    this.severity = severity;
    this.message = message;
    this.timestamp = timestamp;
    this.metricValue = metricValue;
  }
}

class AlertEngine {
  constructor() {
    this.rules = [];
    this._setupDefaultRules();
  }

  _setupDefaultRules() {
    // Rule 1: High error rate
    this.addRule(new AlertRule(
      "high_error_rate",
      (m) => { try { return (m.errorRate ?? m.error_rate ?? 0) > 5; } catch { return false; } },
      "critical",
      "Error rate is {value}% (threshold: 5%)",
      (m) => (m.errorRate ?? m.error_rate ?? 0).toFixed(1),
    ));

    // Rule 2: High latency p95
    this.addRule(new AlertRule(
      "high_latency_p95",
      (m) => { try { return (m.latency?.p95 ?? 0) > 10000; } catch { return false; } },
      "warning",
      "P95 latency is {value}ms (threshold: 10,000ms)",
      (m) => Math.round(m.latency?.p95 ?? 0),
    ));

    // Rule 3: High cost
    this.addRule(new AlertRule(
      "high_cost",
      (m) => {
        try {
          const cph = m.tokens?.costPerHour ?? m.tokens?.cost_per_hour ?? 0;
          return cph > 1.0;
        } catch { return false; }
      },
      "warning",
      "Cost rate is ${value}/hr (threshold: $1.00/hr)",
      (m) => (m.tokens?.costPerHour ?? m.tokens?.cost_per_hour ?? 0).toFixed(2),
    ));

    // Rule 4: Tool failure rate
    this.addRule(new AlertRule(
      "tool_failure_rate",
      (m) => {
        try {
          const tools = m.tools ?? {};
          for (const stats of Object.values(tools)) {
            if ((stats.failureRate ?? stats.failure_rate ?? 0) > 10) return true;
          }
          return false;
        } catch { return false; }
      },
      "critical",
      "Tool failure rate is {value}% (threshold: 10%)",
      (m) => {
        let max = 0;
        for (const stats of Object.values(m.tools ?? {})) {
          const rate = stats.failureRate ?? stats.failure_rate ?? 0;
          if (rate > max) max = rate;
        }
        return max.toFixed(1);
      },
    ));

    // Rule 5: Stale
    this.addRule(new AlertRule(
      "stale_no_requests",
      (m) => {
        try {
          const rpm = m.throughputRpm ?? m.throughput_rpm ?? 0;
          const uptime = m.uptimeSeconds ?? m.uptime_seconds ?? 0;
          return rpm === 0 && uptime > 300;
        } catch { return false; }
      },
      "info",
      "No requests in last 5 minutes (uptime: {value}s)",
      (m) => Math.round(m.uptimeSeconds ?? m.uptime_seconds ?? 0),
    ));
  }

  addRule(rule) {
    this.rules.push(rule);
  }

  evaluate(metrics) {
    const triggered = [];

    for (const rule of this.rules) {
      try {
        if (rule.conditionFn(metrics)) {
          const value = rule.valueFn ? rule.valueFn(metrics) : null;
          const message = value !== null
            ? rule.messageTemplate.replace("{value}", value)
            : rule.messageTemplate;
          triggered.push(new Alert(
            rule.name,
            rule.severity,
            message,
            Date.now(),
            value,
          ));
        }
      } catch {
        // Skip bad rules
      }
    }

    return triggered;
  }
}

// =============================================================================
// SELF-TEST
// =============================================================================

function selfTest() {
  console.log("=".repeat(60));
  console.log("M20 Alert Engine — Self-Test");
  console.log("=".repeat(60));

  const engine = new AlertEngine();

  const healthyMetrics = {
    requestCount: 500,
    errorRate: 1.2,
    latency: { p50: 200, p75: 400, p90: 800, p95: 1500, p99: 3000 },
    tokens: { totalInput: 500000, totalOutput: 250000, costEstimate: 0.5, costPerHour: 0.25 },
    tools: {
      search_filings: { calls: 300, failures: 5, failureRate: 1.7, avgDurationMs: 45 },
      get_details: { calls: 200, failures: 2, failureRate: 1.0, avgDurationMs: 12 },
    },
    throughputRpm: 8.3,
    uptimeSeconds: 3600,
  };

  let alerts = engine.evaluate(healthyMetrics);
  console.log(`\nHealthy metrics -> ${alerts.length} alerts`);
  if (alerts.length !== 0) throw new Error(`Expected 0 alerts, got ${alerts.length}`);
  console.log("  ✅ No alerts — PASS");

  const degradedMetrics = {
    requestCount: 500,
    errorRate: 8.5,
    latency: { p50: 2000, p75: 5000, p90: 9000, p95: 12000, p99: 18000 },
    tokens: { totalInput: 5000000, totalOutput: 2500000, costEstimate: 52.5, costPerHour: 2.1 },
    tools: {
      search_filings: { calls: 300, failures: 45, failureRate: 15.0, avgDurationMs: 145 },
      get_details: { calls: 200, failures: 2, failureRate: 1.0, avgDurationMs: 12 },
    },
    throughputRpm: 8.3,
    uptimeSeconds: 3600,
  };

  alerts = engine.evaluate(degradedMetrics);
  console.log(`\nDegraded metrics -> ${alerts.length} alerts`);
  const icons = { critical: "🔴", warning: "🟡", info: "🔵" };
  for (const alert of alerts) {
    console.log(`  ${icons[alert.severity]} [${alert.severity.toUpperCase()}] ${alert.ruleName}: ${alert.message}`);
  }
  if (alerts.length < 3) throw new Error(`Expected >= 3 alerts, got ${alerts.length}`);
  if (!alerts.some((a) => a.severity === "critical")) throw new Error("Expected critical alert");
  console.log("  ✅ Multiple alerts with correct severities — PASS");

  const staleMetrics = {
    requestCount: 0,
    errorRate: 0,
    latency: { p50: 0, p75: 0, p90: 0, p95: 0, p99: 0 },
    tokens: { totalInput: 0, totalOutput: 0, costEstimate: 0, costPerHour: 0 },
    tools: {},
    throughputRpm: 0,
    uptimeSeconds: 600,
  };

  alerts = engine.evaluate(staleMetrics);
  console.log(`\nStale metrics -> ${alerts.length} alerts`);
  for (const alert of alerts) {
    console.log(`  ${icons[alert.severity]} [${alert.severity.toUpperCase()}] ${alert.ruleName}: ${alert.message}`);
  }
  const staleAlerts = alerts.filter((a) => a.ruleName === "stale_no_requests");
  if (staleAlerts.length < 1) throw new Error("Expected stale alert");
  console.log("  ✅ Stale detection works — PASS");

  console.log(`\n${"=".repeat(60)}`);
  console.log("Self-test complete!");
  console.log("=".repeat(60));
}

export { AlertEngine, AlertRule, Alert };

const isMain = process.argv[1] && process.argv[1].endsWith("alert_engine.js");
if (isMain) {
  selfTest();
}
